daffodil-runtime2/src/test/resources/org/apache/daffodil/runtime2/variablelen.dfdl.xsd

```diff
@@ -0,0 +1,121 @@
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+
+<xs:schema
+ elementFormDefault="qualified"
+ xmlns:dfdl="http://www.ogf.org/dfdl/dfdl-1.0/"
+ xmlns:xs="http://www.w3.org/2001/XMLSchema">
+
+  <!-- Binary data format bindings -->
+
+  <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd"/>
+  <xs:annotation>
+    <xs:appinfo source="http://www.ogf.org/dfdl/">
+      <dfdl:format
+        ref="GeneralFormat"
+        binaryBooleanFalseRep="0"
+        binaryBooleanTrueRep="1"
+        fillByte="%NUL;"
+        prefixIncludesPrefixLength="no"
+        representation="binary"/>
+    </xs:appinfo>
+  </xs:annotation>
+
+  <!-- Types for testing -->
+
+  <!-- Unsuitable since variable length array must allow 0 to 16 numbers -->
+  <xs:complexType name="fixedType">
+    <xs:sequence>
+      <xs:element name="before" type="xs:unsignedInt" />
+      <xs:element name="variablelen_size" type="xs:unsignedInt" />
+      <xs:element name="variablelen" type="xs:unsignedInt"
+                  minOccurs="2" maxOccurs="2"
+                  dfdl:occursCountKind="fixed" />
+      <!-- Yes, can have "after" element after "fixed" array -->
+      <xs:element name="after" type="xs:unsignedInt"
+                  minOccurs="2" maxOccurs="2"
+                  dfdl:occursCountKind="fixed" />
+    </xs:sequence>
+  </xs:complexType>
+
+  <!-- Unsuitable since "implicit" works only at end of binary data -->
+  <xs:complexType name="implicitType">
+    <xs:sequence>
+      <xs:element name="before" type="xs:unsignedInt" />
+      <xs:element name="variablelen_size" type="xs:unsignedInt" />
+      <xs:element name="variablelen" type="xs:unsignedInt"
+                  minOccurs="0" maxOccurs="16"
+                  dfdl:occursCountKind="implicit" />
+      <!-- No, cannot have "after" element after "implicit" array -->
+    </xs:sequence>
+  </xs:complexType>
+
+  <!-- Unsuitable since "parsed" works only at end of binary data -->
+  <xs:complexType name="parsedType">
+    <xs:sequence>
+      <xs:element name="before" type="xs:unsignedInt" />
+      <xs:element name="variablelen_size" type="xs:unsignedInt" />
+      <xs:element name="variablelen" type="xs:unsignedInt"
+                  minOccurs="0" maxOccurs="16"
+                  dfdl:occursCountKind="parsed" />
+      <!-- No, cannot have "after" element after "parsed" array -->
+    </xs:sequence>
+  </xs:complexType>
+
+  <!-- Suitable, although need to support variable length arrays in C -->
+  <xs:complexType name="expressionType">
+    <xs:sequence>
+      <xs:element name="before" type="xs:unsignedInt" />
+      <xs:element name="variablelen_size" type="xs:unsignedInt" />
+      <xs:element name="variablelen" type="xs:unsignedInt"
+                  minOccurs="0" maxOccurs="16"
```

Comment:
If we encounter DFDL schemas with large or unbounded arrays, then we'll add code to allocate memory from the heap and free it as needed for the information set residing in memory.  We can either say you must use unbounded arrays if you want dynamic arrays or offer a tunable limit above which sufficiently large arrays become dynamic arrays instead of static arrays automatically.